sklearn/metrics/_plot/confusion_matrix.py

```diff
@@ -83,22 +83,38 @@ def plot(self, include_values=True, cmap='viridis',
         n_classes = cm.shape[0]
         self.im_ = ax.imshow(cm, interpolation='nearest', cmap=cmap)
         self.text_ = None
+        if (values_format is None and cm.dtype.kind == "f"):
+            values_format = '.2g'
 
         cmap_min, cmap_max = self.im_.cmap(0), self.im_.cmap(256)
 
         if include_values:
             self.text_ = np.empty_like(cm, dtype=object)
-            if values_format is None:
-                values_format = '.2g'
 
             # print text with appropriate color depending on background
             thresh = (cm.max() + cm.min()) / 2.0
-            for i, j in product(range(n_classes), range(n_classes)):
-                color = cmap_max if cm[i, j] < thresh else cmap_min
-                self.text_[i, j] = ax.text(j, i,
-                                           format(cm[i, j], values_format),
-                                           ha="center", va="center",
-                                           color=color)
+
+            if (values_format is None):
+                for i, j in product(range(n_classes), range(n_classes)):
+
+                    if (cm[i, j] == 0 or np.log10(cm[i, j]) < 7):
+                        values_format = 'd'
+                    else:
+                        values_format = '.2g'
+
+                    color = cmap_max if cm[i, j] < thresh else cmap_min
+
+                    self.text_[i, j] = ax.text(
+                        j, i, format(cm[i, j], values_format),
+                        ha="center", va="center",
+                        color=color)
```

Comment:
These lines are duplicated.

Let's put the `if value_format is None` condition inside of the for loop to avoid duplication.